pyro/infer/inspect.py

```diff
@@ -435,6 +437,31 @@ def generate_graph_specification(
     }
 
 
+def _deep_merge(things: list):
+    if len(things) == 1:
+        return things[0]
+
+    # Recurse into dicts.
+    if isinstance(things[0], dict):
+        result = {}
+        for thing in things:
+            for key, value in thing.items():
+                if key not in result:
+                    result[key] = _deep_merge([t[key] for t in things])
```

Comment:
The logic is a little bit complicated to follow so it would be nice to add a test for `_deep_merge`.